src/cmdline/cmdline_queue.ts

```diff
@@ -0,0 +1,71 @@
+import { type EventBusData } from "../eventBus";
+
+/**
+ * A queue for "batching" cmdline events.
+ *
+ * In the most simple cases, events will simply be passed through and not queued. However, there is a more complicated=
+ * case we must deal with. There is an inherent "race condition" (NB: it's not really a race condition, rather a result
+ * of how the JS event loop works, but it's easiest to think of it as a race) between quickpick hide events and cmdline_* events from nvim. If there is a
+ * cmdline_hide, followed immediately by a cmdline_show, we may call hide() on the quickpick, but onHide will not
+ * be fired until our event handlers have completed execution, leading to bizarre and confusing states. As such,
+ * we "queue" events until we know for sure that the quickpick has bene hidden, and then allow them to be flushed back
+ * to the cmdline_manager.
+ */
+export class CmdlineQueue {
+    private pendingBatches: EventBusData<"redraw">[][] = [];
+    private needFlush: boolean = false;
+    private level: number | null = null;
```

Comment:
Are we tracking level in two places now?

Reply:
Unfortunately, yes. This gave me a bit of an ick, but they're acutally used for different purposes; this will not always be in sync with the `level` on the cmdline manager itself. I could probably name this something better.

Reply:
Hopefully the new naming makes it a bit clearer.